src/lang/std/artifactGraph.ts

```diff
@@ -0,0 +1,645 @@
+}
+interface ExtrusionArtifact {
+  type: 'extrusion'
+  pathId: string
+  surfs: Array<_WallArtifact | _CapArtifact>
+  edges: Array<string>
+  codeRef: CommonCommandProperties
+}
+
+interface _WallArtifact {
+  type: 'wall'
+  segId: string
+  edgeCutEdgeIds: Array<string>
+  extrusionId: string
+  pathIds: Array<string>
+}
+interface _CapArtifact {
+  type: 'cap'
+  subType: 'start' | 'end'
+  edgeCutEdgeIds: Array<string>
+  extrusionId: string
+  pathIds: Array<string>
+}
+
+interface _ExtrudeEdge {
+  type: 'extrudeEdge'
+  segId: string
+  extrusionId: string
+  edgeId: string
+}
+
+/** A edgeCut is a more generic term for both fillet or chamfer */
+interface _EdgeCut {
+  type: 'edgeCut'
+  subType: 'fillet' | 'chamfer'
+  consumedEdgeId: string
+  edgeIds: Array<string>
+  surfId: string
+  codeRef: CommonCommandProperties
+}
+
+interface EdgeCutEdge {
+  type: 'edgeCutEdge'
+  edgeCutId: string
+  surfId: string
+}
+
+export type Artifact =
+  | _PlaneArtifact
+  | _PathArtifact
+  | _SegmentArtifact
+  | _ExtrusionArtifact
+  | _WallArtifact
+  | _CapArtifact
+  | _ExtrudeEdge
+  | _EdgeCut
+  | EdgeCutEdge
+  | _solid2D
+
+export type ArtifactGraph = Map<string, Artifact>
+
+export type EngineCommand = Models['WebSocketRequest_type']
+
+type OkWebSocketResponseData = Models['OkWebSocketResponseData_type']
+
+export interface ResponseMap {
+  [commandId: string]: OkWebSocketResponseData
+}
+export interface OrderedCommand {
+  command: EngineCommand
+  range: SourceRange
+}
+
+export function createArtifactGraph({
+  orderedCommands,
+  responseMap,
+  ast,
+}: {
+  orderedCommands: Array<OrderedCommand>
+  responseMap: ResponseMap
+  ast: Program
+}) {
+  const myMap = new Map<string, Artifact>()
+  let currentPlaneId = ''
+
+  const getArtifact = (id: string) => myMap.get(id)
+  orderedCommands.forEach((orderedCommand) => {
+    if (orderedCommand.command?.type === 'modeling_cmd_req') {
+      if (orderedCommand.command.cmd.type === 'enable_sketch_mode') {
+        currentPlaneId = orderedCommand.command.cmd.entity_id
+      }
+      if (orderedCommand.command.cmd.type === 'sketch_mode_disable') {
+        currentPlaneId = ''
+      }
+    }
+    const modArr = getArtifactsToUpdate({
+      orderedCommand,
+      responseMap,
+      getArtifact,
+      currentPlaneId,
+      ast,
+    })
+    modArr.forEach(({ id, artifact }) => {
+      const oldArtifact = myMap.get(id)
+      if (oldArtifact?.type !== artifact.type) {
+        myMap.set(id, artifact)
+        return
+      }
+      const mergedArtifact = { ...oldArtifact, ...artifact }
+      Object.entries(artifact).forEach(([propName, value]) => {
+        const otherValue = (oldArtifact as any)[propName]
+        if (Array.isArray(value) && Array.isArray(otherValue)) {
+          ;(mergedArtifact as any)[propName] = [
+            ...new Set([...otherValue, ...value]),
+          ]
+        }
+      })
+      myMap.set(id, mergedArtifact)
+    })
```

Comment:
If anyone knows a better way of typing this I'm all ears.

Reply:
I might pull out the merging to a separate function `mergeArtifacts(oldArtifact, newArtifact): Artifact`. But otherwise, I think it's fine. Why, what were you thinking?

The current implementation is pretty specific when you think about all the potential edge cases of a general object merge. Like what if the old value is an array and the new one isn't, or vice versa? I'm guessing we just don't care because our artifacts have stable types based on the `type` field.

Reply:
That's fair, did that in bafd5db0c6dd453776ba564b3bb654bbebd84a4e

Changed the typing a little but it's just different not really better.

Didn't really have anything in mind, would be nice to have it type narrow on `oldArtifact.type === newArtifact.type` But that's asking the static analyser to do a lot.

I had thought I could have a switch statement to cover each artifact type, but it seems like bloat just to make TS happy when all I'm trying to do is merge arrays.